File: Data Structures & Algorithms/meeting-schedule-ii/submission-0.java
/**
 * Definition of Interval:
 * public class Interval {
 *     public int start, end;
 *     public Interval(int start, int end) {
 *         this.start = start;
 *         this.end = end;
 *     }
 * }
 */

class Solution {
    public int minMeetingRooms(List<Interval> intervals) {
        int n=intervals.size();
        int[] start=new int[n];
        int[] end=new int[n];
        for(int i=0;i<n;i++){
            start[i]=intervals.get(i).start;
            end[i]=intervals.get(i).end;
        }
        Arrays.sort(start);
        Arrays.sort(end);
        int res=0,c=0,s=0,e=0;
        while(s<n){
            if(start[s]<end[e]){
                s++;
                c++;
            }else{
                e++;
                c--;
            }
            res=Math.max(res,c);
        }

return res;
    }
}

File: Data Structures & Algorithms/merge-intervals/submission-0.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if not intervals:
            return []

        # 1) sort by start time
        intervals.sort(key=lambda x: x[0])

        res = []
        cur_start, cur_end = intervals[0]

        # 2) merge by scanning
        for start, end in intervals[1:]:
            # overlap case: next interval starts before current ends
            if start <= cur_end:
                cur_end = max(cur_end, end)   # extend current interval
            else:
                res.append([cur_start, cur_end])  # close current interval
                cur_start, cur_end = start, end   # start new interval

        # 3) add the last interval
        res.append([cur_start, cur_end])
        return res
